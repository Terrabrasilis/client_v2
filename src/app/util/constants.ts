/**
 * This class is responsible to store all global variables to use in entire application and not duplicate code
 */
export class Constants {
    public static get TERRABRASILIS_MAPS_GWC(): string { 
        return "http://terrabrasilis.info/fip-service/gwc/service/wms"; 
    };

    public static get TERRABRASILIS_MAPS_WMS(): string { 
        return "http://terrabrasilis.info/fip-service/wms"; 
    };

    public static get FIPCERRADO_OPERACAO(): string { 
        return "http://fipcerrado.dpi.inpe.br:8080/fipcerrado-geoserver/terraamazon/wms"; 
    };

    public static get FEATURE_INFO_PARAMS(): string {
        let host = "{0}"
        let service = "/wms?SERVICE=WMS";
        let version = "&VERSION=1.1.1";
        let request = "&REQUEST=GetFeatureInfo";
        let layers = "&LAYERS={1}";
        let query = "&QUERY_LAYERS={2}";
        let style = "&STYLES=";
        let bbox = "&BBOX={3}";
        let count = "&FEATURE_COUNT=";
        let width = "&WIDTH={4}";
        let heigth = "&HEIGHT={5}";
        let format = "&FORMAT=";
        let infoFormat = "&INFO_FORMAT={6}";
        let srs = "&SRS=EPSG:4326";
        let x = "&X={7}";
        let y = "&Y={8}";

        return host + service + version + request + layers + query + style + bbox + count + width + heigth + format + infoFormat + srs + x + y;
    }

    public static get PROXY_OGC(): string { 
        return "http://terrabrasilis.dpi.inpe.br/proxy?url="; 
    };

    public static get DASHBOARD_API_HOST(): string {
        return "http://terrabrasilis.dpi.inpe.br/dashboard/api/v1/redis-cli/";
    };

    public static get DASHBOARD_BIOMES_COLORS(): string[] {
        let listColors: string[] = ["#339966", "#ffcc00", "#ffff99", "#ff6600", "#33cccc", "#66ff33"];
        return listColors;
    };

    public static get DASHBOARD_BIOMES_NAMES(): string[] {
        let listNames: string[] = ["amazon", "atlantic", "caatinga", "cerrado", "pampa", "pantanal", "legal_amazon"];
        return listNames;
    };

    public static get DASHBOARD_CERRADO_DUPLICATED_YEARS(): number[] {
        let duplicatedYears: number[] = [2002, 2004, 2006, 2008, 2010, 2012];
        return duplicatedYears;
    };        

    public static get DASHBOARD_CERRADO_MAINTAINABLE_YEARS(): number[] {
        let maitanableYears: number[] = [2013, 2014, 2015, 2016, 2017, 2018];
        return maitanableYears;
    };

    public static get MAP_LEGEND_COLORS(): any[] {
        return ['#ffffcc', '#ffeda0', '#fed976', '#feb24c', '#fd8d3c', '#fc4e2a', '#e31a1c', '#bd0026', '#800026'];
    }

    public static get TERRABRASILIS_API_HOST(): string {
        return "http://terrabrasilis.dpi.inpe.br/terrabrasilis/api/v1/";
    };    

    public static get MAP_LEGEND_GRADES(): number {
        return 8;
    };  
    
    public static get DASHBOARD_CERRADO_STATES(): any[] {
        return ['MATO GROSSO', 'MARANHÃO', 'PIAUÍ', 'BAHIA', 'MATO GROSSO DO SUL', 'GOIÁS', 'MINAS GERAIS', 'SÃO PAULO', 'PARANÁ', 'TOCANTINS', 'DISTRITO FEDERAL'];
    };  

    public static get DASHBOARD_AMAZON_STATES(): any[] {
        return ['PARÁ', 'AMAZONAS', 'RORAIMA', 'ACRE', 'MATO GROSSO', 'RONDÔNIA', 'AMAPÁ', 'MARANHÃO', 'TOCANTINS'];
    };  

    public static get DASHBOARD_LEGAL_AMAZON_STATES(): any[] {
        return ['PARÁ', 'AMAZONAS', 'RORAIMA', 'ACRE', 'MATO GROSSO', 'RONDÔNIA', 'AMAPÁ', 'MARANHÃO', 'TOCANTINS'];
    };  
    
    public static get TERRABRASILIS_BUSINESS_API_HOST(): string {
        return (process.env.ENV == 'production')?("http://terrabrasilis.dpi.inpe.br/business/api/v1/"):("http://terrabrasilis.dpi.inpe.br/business/api/v1/");
    };
}
