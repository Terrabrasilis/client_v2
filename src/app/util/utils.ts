import { includes, filter, forEach, kebabCase, get } from 'lodash';

export const Utils = {

  genericSearch (arrayToBeSearched = [], searchText = '') {
    const searchWords = this.removeSpecialCharactersAndCase(searchText)
    const elementsFound = filter(arrayToBeSearched, (element) => {
      let isValid = true
      const arrayElement = this.removeSpecialCharactersAndCase(element)
      forEach(searchWords, (searchedWord) => {
        if (!includes(arrayElement, searchedWord)) isValid = false
      })
      return isValid
    })

    return elementsFound
  },

  removeSpecialCharactersAndCase (word = '') {
    return kebabCase(word.toLowerCase()).split('-').join(' ')
  }

}
