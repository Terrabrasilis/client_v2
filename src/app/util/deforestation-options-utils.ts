/**
 * This class is responsible to expose auxiliary functions within deforestation options code
 */    

import * as Terrabrasilis from "terrabrasilis-api";
import { Constants } from './constants';
import * as dc from "dc";
import * as d3 from "d3";

export class DeforestationOptionsUtils {

  public static setLoiNamesDownload(loi:any, loinames:any, checkedLoiNames:any) {
        
    loi.loinames.forEach(function(loiname:any) {  
      var [mun, state] = loiname.loiname.split("_");
      if (checkedLoiNames.indexOf(state) > -1)
        loinames[loiname.gid] = [mun, state];
    });

    return;

  }

  public static setLoiNames(loi:any, self:any) {
    
    loi.loinames.forEach(function(loiname:any) {  
      self.loiNames[loiname.gid] = loiname.loiname;
      self.loiNamesObject.push({key:loiname.gid,value:loiname.loiname});
    });

  }

  public static setLoiNamesSplit(loi:any, self:any) {
    
    loi.loinames.forEach(function(loiname:any) {  
      var value = loiname.loiname.split("_")[0];
      self.loiNames[loiname.gid] = value;
      self.loiNamesObject.push({key:loiname.gid,value:value});
    });

  }

  public static dataWranglingRates(dataJson:any) {
    
    var all:any[] = [];

    dataJson["periods"].forEach(function(period:any) {
      period.features.forEach(function(feature:any) {       
        var year = period.endDate.year;          
        var area = feature.areas.filter((area:any) => area.type == 1).map(function(e:any) { return e.area; })[0];
        var filteredArea = feature.areas.filter((area:any) => area.type == 2).map(function(e:any) { return e.area; })[0];
        all.push({ 
          endDate: year,
          loi: feature.loi,
          loiName: feature.loiname,
          area: area,
          filteredArea: filteredArea
        });
      });
    });
        
    return all;
  }

  public static dataWranglingIncrements(dataJson:any, biome:any) {

    var all:any[] = [];

    if(biome == "cerrado") {
      dataJson["periods"].forEach(function(period:any) {
        period.features.forEach(function(feature:any) {       
          var year = period.endDate.year;          
          var area = feature.areas.filter((area:any) => area.type == 2).map(function(e:any) { return e.area; })[0];
          var filteredArea = feature.areas.filter((area:any) => area.type == 1).map(function(e:any) { return e.area; })[0];
          if(Constants.DASHBOARD_CERRADO_DUPLICATED_YEARS.indexOf(year) > -1) {
            all.push({ 
              endDate: year,
              loi: feature.loi,
              loiName: feature.loiname,
              area: area*0.5,
              filteredArea: filteredArea*0.5
            });
            all.push({
              endDate: year-1,
              loi: feature.loi,
              loiName: feature.loiname,
              area: area*0.5,
              filteredArea: filteredArea*0.5
            });
          } else  if(Constants.DASHBOARD_CERRADO_MAINTAINABLE_YEARS.indexOf(year) > -1) {
                    all.push({ 
                              endDate: year, 
                              loi: feature.loi, 
                              loiName: feature.loiname, 
                              area: area,
                              filteredArea: filteredArea
                            });
                  }
        });
      });
    }
    else if (biome == "amazon" || biome == "legal_amazon") {
      dataJson["periods"].forEach(function(period:any) {
        period.features.forEach(function(feature:any) {       
          var year = period.endDate.year;          
          var area = feature.areas.filter((area:any) => area.type == 2).map(function(e:any) { return e.area; })[0];
          var filteredArea = feature.areas.filter((area:any) => area.type == 1).map(function(e:any) { return e.area; })[0];
          all.push({
                      endDate: year, 
                      loi: feature.loi, 
                      loiName: feature.loiname, 
                      area: area,
                      filteredArea: filteredArea
                  });
        });
      });
    }

    return all;

  }
    
  public static getloiNamesByLoi(arr:any, loi:any):Array<number> {      
    
    return arr.filter(
      (filteredloi:any) => { 
        return filteredloi.loi === loi
      }
    )
    .map(
      (e:any) => { 
        return {"key":e["gid"], "value":e["loiname"]};
      }
    );

  };

  public static style(feature:any) {
      
      return {
          fillColor: Terrabrasilis.getColor(feature.properties.density),
          weight: 2,
          opacity: 1,
          color: 'white',
          dashArray: '3',
          fillOpacity: 0.7
      };
  
  };

  // format Number
  public static formatNumber(num:number) {
    
    if (num >= 1000000000) {
        return (num / 1000000000).toFixed(1).replace(/\.0$/, '') + 'G';
    } else if (num >= 1000000) {
        return (num / 1000000).toFixed(1).replace(/\.0$/, '') + 'M';
    } else if (num >= 1000) {
        return (num / 1000).toFixed(1).replace(/\.0$/, '') + 'K';
    }        

    return (num).toFixed(1).replace(/\.0$/, '');

  };

  public static get MAP_LEGEND_GRADES(): number {
      return 8;
  };    

  public static renderGraph(id:any, listCharts:any, transition:any, loiNames:any, loi:any, type: any):void {
    
    var width = $('#'+id).width(); // get width from parent div
    var height = $('#'+id).height(); // get height from parent div
  
    var result = listCharts.get(id); // get result from list charts
        
    if (id == "bar-chart") {      
        var barChartLegend = Constants.DASHBOARD_LEGEND_WIDTH_BAR_CHART;    
        
        result.width(width-barChartLegend) // update width
              .height(height) // update height
              .transitionDuration(transition); // update transitions

        result.x(d3.scaleBand().rangeRound([0, width]).paddingInner(0.05))
              .xUnits(dc.units.ordinal);  

        if (type != "rates")         
          result.legend(dc.legend().x(width-barChartLegend).y(5).itemHeight(13).gap(4).legendText(function(d:any) { return d.name; }));        

        result.on("renderlet.a",function (chart:any) {
          // rotate x-axis labels
          chart.selectAll('g.x text')
            .attr('transform', 'translate(-10,10) rotate(315)');
          $("#bar-chart > svg").attr("width", width);
        });

    } else if (id == "series-chart") {

              var seriesChartLegend = Constants.DASHBOARD_LEGEND_WIDTH_SERIES_CHART.get(loi);
              
              result.width(width-seriesChartLegend) // update width
                    .height(height) // update height
                    .transitionDuration(transition); // update transitions
              
              result.legend(dc.legend().x(width-seriesChartLegend).y(10).itemHeight(13).gap(5).legendText(function(d:any) { 
                return loiNames[d.name];
              }));

              result.on("renderlet.a",function (chart:any) {
                // rotate x-axis labels
                chart.selectAll('g.x text')
                  .attr('transform', 'translate(-10,10) rotate(315)');
                $("#series-chart > svg").attr("width", width);
              });

            }
            else if (id == "row-chart") {
                result.width(width*.95) // update width
                .height(height*.95) // update height
                .transitionDuration(transition); // update transitions

                $('.search-loi').width(0.8*width);
            }

    (function(j, result){
      setTimeout(() => {
        result.render();  
        Terrabrasilis.disableLoading(id);               
        Terrabrasilis.disableLoading("#loi-chart");      
      },100 * j);
    })(2, result);    

  }

  public static renderMap():void {
    Terrabrasilis.enableInvalidateSize();
    
  }

  public static render(item:any, listCharts:any, transition:any, loiNames:any, loi:any, type:any):void {
    if (item != "loi-chart")        
      this.renderGraph(item, listCharts, transition, loiNames, loi, type);
    else
      this.renderMap();
  }
    
}