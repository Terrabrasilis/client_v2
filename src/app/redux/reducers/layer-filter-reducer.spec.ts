import {
  actions,
  featureKey,
  reducer,
  onSetFilterAction,
  State
} from "./layer-filter-reducer";
import { log } from "util";

describe("LayerFilterReducer", () => {
  describe("ACTIONS:", () => {
    it('feature key should be "layerFilter"', () => {
      expect(featureKey).toBe("layerFilter");
    });

    it("setInitialDate should be a function", () => {
      expect(actions.setFilterPropsForObject).toBeInstanceOf(Function);
    });
  });

  describe("REDUCERS", () => {
    it("should return the default state", () => {
      const result = reducer(undefined, {} as any);
      expect(result).toMatchSnapshot();
    });

    describe("onSetFilterAction", () => {
      it("should push a new object into the array", () => {
        const results = onSetFilterAction(undefined, { id: 1 });
        expect(results).toStrictEqual({ filters: [{ id: 1 }] });
      });
      it("should merge with an existent object", () => {
        const dateMock = new Date(2019, 6, 6);
        const initialDate = new Date(2016, 0, 1);
        const finalDate = new Date(2016, 0, 1);
        const initialState: State = {
          filters: [{ id: 1, initialDate, finalDate }]
        };
        const results = onSetFilterAction(initialState, {
          id: 2,
          initialDate: dateMock
        });
        expect(results).toStrictEqual({
          filters: [
            { id: 1, initialDate, finalDate },
            { id: 2, initialDate: dateMock }
          ]
        });
      });

      it("should update an existent object", () => {
        const dateMock = new Date(2019, 6, 6);
        const initialDate = new Date(2016, 0, 1);
        const finalDate = new Date(2016, 0, 1);
        const initialState: State = {
          filters: [{ id: 1, initialDate, finalDate }]
        };
        const results = onSetFilterAction(initialState, {
          id: 1,
          initialDate: dateMock
        });
        expect(results).toStrictEqual({
          filters: [{ id: 1, initialDate: dateMock, finalDate }]
        });
      });
    });
  });

  describe("SELECTORS", () => {});
});
