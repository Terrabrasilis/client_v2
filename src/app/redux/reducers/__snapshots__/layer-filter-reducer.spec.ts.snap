// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerFilterReducer REDUCERS should return the default state 1`] = `
Object {
  "filters": Array [],
}
`;
